go-controller/pkg/ovn/egressip.go

```diff
@@ -0,0 +1,542 @@
+	oc.eIPAllocatorMutex.Lock()
+	defer oc.eIPAllocatorMutex.Unlock()
+
+	if len(oc.eIPAllocator) == 0 {
+		oc.egressAssignmentRetry.Store(eIP, true)
+		return nil, fmt.Errorf("no assignable nodes")
+	}
+
+	assignments := []egressipv1.EgressIPStatus{}
+	eNodes, existingAllocations := oc.getSortedEgressData()
+	klog.V(5).Infof("Current assignments are: %+v", existingAllocations)
+	for _, egressIP := range eIP.Spec.EgressIPs {
+		klog.V(5).Infof("Will attempt assignment for egress IP: %s", egressIP)
+		eIPC := net.ParseIP(egressIP)
+		if eIPC == nil {
+			klog.Errorf("Unable to parse provided EgressIP: %s, invalid", egressIP)
+			continue
+		}
+		for i := 0; i < len(eNodes); i++ {
+			klog.V(5).Infof("Attempting assignment on egress node: %+v", eNodes[i])
+			if _, exists := existingAllocations[eIPC.String()]; !exists {
+				if !eNodes[i].tainted {
+					if (utilnet.IsIPv6(eIPC) && eNodes[i].v6Subnet != nil && eNodes[i].v6Subnet.Contains(eIPC)) ||
+						(!utilnet.IsIPv6(eIPC) && eNodes[i].v4Subnet != nil && eNodes[i].v4Subnet.Contains(eIPC)) {
+						eNodes[i].tainted, oc.eIPAllocator[eNodes[i].name].allocations[eIPC.String()] = true, true
+						assignments = append(assignments, egressipv1.EgressIPStatus{
+							EgressIP: eIPC.String(),
+							Node:     eNodes[i].name,
+						})
+						klog.V(5).Infof("Successful assignment of egress IP: %s on node: %+v", egressIP, eNodes[i])
+						break
+					}
+				}
+			}
+		}
+	}
+	if len(assignments) == 0 {
+		return nil, fmt.Errorf("no matching host found")
+	}
+	return assignments, nil
+}
+
+func (oc *Controller) releaseEgressIPs(eIP *egressipv1.EgressIP) error {
+	oc.eIPAllocatorMutex.Lock()
+	defer oc.eIPAllocatorMutex.Unlock()
+	for _, status := range eIP.Status {
+		klog.V(5).Infof("Releasing egress IP assignment: %s", status.EgressIP)
+		if node, exists := oc.eIPAllocator[status.Node]; exists {
+			delete(node.allocations, status.EgressIP)
+		}
+		klog.V(5).Infof("Remaining allocations on node are: %+v", oc.eIPAllocator[status.Node])
+	}
+	return nil
+}
+
+func (oc *Controller) getSortedEgressData() ([]eNode, map[string]bool) {
+	nodes := []eNode{}
+	allAllocations := make(map[string]bool)
+	for _, eNode := range oc.eIPAllocator {
+		nodes = append(nodes, *eNode)
+		for ip := range eNode.allocations {
+			allAllocations[ip] = true
+		}
+	}
+	sort.Slice(nodes, func(i, j int) bool {
+		return len(nodes[i].allocations) < len(nodes[j].allocations)
+	})
+	return nodes, allAllocations
+}
+
+func getEgressIPKey(eIP *egressipv1.EgressIP) string {
+	return eIP.Name
+}
+
+func getPodKey(pod *kapi.Pod) string {
+	return fmt.Sprintf("%s_%s", pod.Namespace, pod.Name)
+}
+
+type egressIPMode struct {
+	ovnNBClient    goovn.Client
+	podRetry       sync.Map
+	gatewayIPCache sync.Map
+}
+
+type gatewayRouter struct {
+	ipV4 net.IP
+	ipV6 net.IP
+}
+
+func (e *egressIPMode) getGatewayRouterIP(node string, wantsIPv6 bool) (net.IP, error) {
```

Comment:
I think you are getting the pod facing GR IP here right? for reroute the pod traffic to the right GW? It would be good to clarify this, preferably with a better function name.

Reply:
I thought `getGatewayRouterIP` was clear enough, would `getPodFacingGatewayRouterIP` be better?

Reply:
yeah or getPodNetworkGatewayRouterIP

Reply:
Will call it `getGatewayRouterJoinIP`